### DISCUSSION TITLE
Store and replay video streams

### DISCUSSION THREAD

**Discussion by twdragon** [ros-drivers/usb_cam#251] [Ideas]
The proposal is presented in #208 @knorth55 @flynneva what do you think?  For me, this looks like a redundant feature for the project.

**Comment by knorth55** [comment]
storing function is reasonable to add in usb_cam.
but replaying should be different or use other existing nodes

**Reply by flynneva** [reply]
@twdragon @knorth55 i guess I don't really understand the request.

ROS now has support for compressed image formats directly within the normal sensor msgs Image message definition....so why not just use those and use rosbag2 to record / replay?

What's the benefit of writing our own recording / replay framework?

**Reply by twdragon** [reply]
@flynneva I completely support your opinion, we have already a `rosbag` framework to store and replay the message streams. To implement the store/replay feature in `usb_cam` we actually should write a `libavcodec`-based recorder within the node like in [this example](https://git.ffmpeg.org/gitweb/ffmpeg.git/blob/refs/heads/master:/doc/examples/encode_video.c). I do not see the reason to do this long work in this project. @knorth55 could you please explain a bit about how would you try to store the stream and what instruments you suggest to use?

**Reply by knorth55** [reply]
we use image transport system, and in the system, we first decompress image to create image topic and again compress in the imaget transport plugin.
i thought some people think that it is useless decompression and compression, when you subscribe the compressed image topic and store it in some file.
so i thought the idea storing image directly in compressed format sounds reasonable to me.
but im not a strong follower of the idea because i dont encounter the serious issue on that bottleneck.

**Reply by knorth55** [reply]
for replaying function, there are several ways to publish topic from video data, so i think we dont need to reimplement it.

**Reply by twdragon** [reply]
@knorth55 it seems implementable only if the compressed stream from the camera matches the format of the stream generated by `image_transport` which is almost always not true. The way to eliminate the compression bottleneck is to implement the hardware-accelerated decoder, in my opinion. This will prevent us to invent the wheel, at least our own storage archive format.

**Comment by twdragon** [comment]
Guys, we have agreed, I think. So, we can close both this discussion and the related PR. Thank you for the discussion!
